perception/lidar_transfusion/lib/preprocess/pointcloud_densification.cpp

```diff
@@ -90,7 +92,19 @@ void PointCloudDensification::enqueue(
 {
   affine_world2current_ = affine_world2current;
   current_timestamp_ = rclcpp::Time(msg.header.stamp).seconds();
-  pointcloud_cache_.emplace_front(PointCloudWithTransform{msg, affine_world2current.inverse()});
+
+  assert(sizeof(uint8_t) * msg.width * msg.height * msg.point_step % sizeof(float) == 1);
+  auto points_d = cuda::make_unique<float[]>(
+    sizeof(uint8_t) * msg.width * msg.height * msg.point_step / sizeof(float));
+
+  CHECK_CUDA_ERROR(cudaMemcpyAsync(
```

Comment:
This is a question rather than a review comment that requires changes from your part.
I had understood that async did not really work if the memory was not pinned? Did I misunderstand that part?

Reply:
@amadeuszsz 
Talking with Hirabayashi-san, we kind of got the impression that the operation is not fully parallelized (pageable to pinned may not be async and then the pinned to device is async). If you have any further information, please tell us, otherwise please just close the comment

Reply:
> pageable to pinned may not be async and then the pinned to device is async

Right, but docs are kind of unclear - how to interpret "_may_" in "_the driver may synchronize with the stream_" [here](https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-async)? There was an update between [12.0.0](https://docs.nvidia.com/cuda/archive/12.0.0/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-async) and [12.1.0](https://docs.nvidia.com/cuda/archive/12.1.0/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-async) documentation (we target 12.2.0), the old one is unambiguous and I'm not sure if it's only text refactor or there was some changes under the hood. Anyway, the API call type (e.g. `cudaMemcpyAsync` used for sync opertation) is not that important since CUDA will force sync whenever it's needed. For clarity, I added stream synchronization each time we access the device memory. I didn't observe any overhead and it will return error if the memory was corrupted in previous operation -> if something wrong happen, users will come with appropriate logs.